File: app/src/main/java/org/mifos/mobilebanking/presenters/ThirdPartyTransferPresenter.java
package org.mifos.mobilebanking.presenters;

import android.content.Context;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.api.DataManager;
import org.mifos.mobilebanking.injection.ApplicationContext;
import org.mifos.mobilebanking.models.AccountOptionAndBeneficiary;
import org.mifos.mobilebanking.models.beneficiary.Beneficiary;
import org.mifos.mobilebanking.models.beneficiary.BeneficiaryDetail;
import org.mifos.mobilebanking.models.payload.AccountDetail;
import org.mifos.mobilebanking.models.templates.account.AccountOption;
import org.mifos.mobilebanking.models.templates.account.AccountOptionsTemplate;
import org.mifos.mobilebanking.presenters.base.BasePresenter;
import org.mifos.mobilebanking.ui.views.ThirdPartyTransferView;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;


/**
 * Created by dilpreet on 21/6/17.
 */

public class ThirdPartyTransferPresenter extends BasePresenter<ThirdPartyTransferView> {

    private DataManager dataManager;
    private CompositeDisposable compositeDisposable;

    /**
     * Initialises the RecentTransactionsPresenter by automatically injecting an instance of
     * {@link DataManager} and {@link Context}.
     *
     * @param dataManager DataManager class that provides access to the data
     *                    via the API.
     * @param context     Context of the view attached to the presenter. In this case
     *                    it is that of an {@link android.support.v7.app.AppCompatActivity}
     */
    @Inject
    public ThirdPartyTransferPresenter(DataManager dataManager,
            @ApplicationContext Context context) {
        super(context);
        this.dataManager = dataManager;
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void attachView(ThirdPartyTransferView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        compositeDisposable.clear();
    }

    /**
     * Fetches {@link AccountOptionsTemplate} and {@link List} of {@link Beneficiary} from server
     * and notifies the view to display them. And in case of any error during fetching the required
     * details it notifies the view.
     */
    public void loadTransferTemplate() {

        checkViewAttached();
        getMvpView().showProgress();

        compositeDisposable.add(Observable.zip(dataManager.getThirdPartyTransferTemplate(),
                dataManager.getBeneficiaryList(),
                new BiFunction<AccountOptionsTemplate, List<Beneficiary>,
                                AccountOptionAndBeneficiary>()
                {
                    @Override
                    public AccountOptionAndBeneficiary apply(AccountOptionsTemplate
                                                                    accountOptionsTemplate,
                                                            List<Beneficiary> beneficiaries) {
                        return new AccountOptionAndBeneficiary(accountOptionsTemplate,
                                beneficiaries);
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<AccountOptionAndBeneficiary>() {
                    @Override
                    public void onComplete() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().hideProgress();
                        getMvpView().showError(context.getString(
                                R.string.error_fetching_third_party_transfer_template));
                    }

                    @Override
                    public void onNext(AccountOptionAndBeneficiary accountOptionAndBeneficiary) {
                        getMvpView().hideProgress();
                        getMvpView().showThirdPartyTransferTemplate(accountOptionAndBeneficiary.
                                getAccountOptionsTemplate());
                        getMvpView().showBeneficiaryList(accountOptionAndBeneficiary.
                                getBeneficiaryList());
                    }
                }));
    }

    /**
     * Retrieving {@link List} of {@code accountNumbers} from {@link List} of {@link AccountOption}
     * @param accountOptions {@link List} of {@link AccountOption}
     * @return Returns {@link List} containing {@code accountNumbers}
     */
    public List<AccountDetail> getAccountNumbersFromAccountOptions(List<AccountOption>
                                                                           accountOptions) {
        final List<AccountDetail> accountNumber = new ArrayList<>();
        Observable.fromIterable(accountOptions)
                .filter(new Predicate<AccountOption>() {
                    @Override
                    public boolean test(AccountOption accountOption) throws Exception {
                        return !accountOption.getAccountType().getCode().equals(context.
                                getString(R.string.account_type_loan));
                    }
                })
                .flatMap(new Function<AccountOption, Observable<AccountDetail>>() {
                    @Override
                    public Observable<AccountDetail> apply(AccountOption accountOption) {
                        return Observable.just(new AccountDetail(accountOption.getAccountNo(),
                                accountOption.getAccountType().getValue()));
                    }
                })
                .subscribe(new Consumer<AccountDetail>() {
                    @Override
                    public void accept(AccountDetail accountDetail) throws Exception {
                        accountNumber.add(accountDetail);

                    }
                });
        return accountNumber;
    }

    /**
     * Retrieving {@link List} of {@code accountNumbers} from {@link List} of {@link Beneficiary}
     * @param beneficiaries {@link List} of {@link Beneficiary}
     * @return Returns {@link List} containing {@code accountNumbers}
     */
    public List<BeneficiaryDetail> getAccountNumbersFromBeneficiaries(final List<Beneficiary>
                                                                              beneficiaries) {
        final List<BeneficiaryDetail> accountNumbers = new ArrayList<>();
        Observable.fromIterable(beneficiaries)
                .flatMap(new Function<Beneficiary, Observable<BeneficiaryDetail>>() {
                    @Override
                    public Observable<BeneficiaryDetail> apply(Beneficiary beneficiary) {
                        return Observable.just(new BeneficiaryDetail(beneficiary.getAccountNumber(),
                                beneficiary.getName()));
                    }
                })
                .subscribe(new Consumer<BeneficiaryDetail>() {
                    @Override
                    public void accept(BeneficiaryDetail beneficiaryDetail) throws Exception {
                        accountNumbers.add(beneficiaryDetail);

                    }
                });
        return accountNumbers;
    }

    /**
     * Searches for a {@link AccountOption} with provided {@code accountNo} from {@link List} of
     * {@link AccountOption} and returns it.
     * @param accountOptions {@link List} of {@link AccountOption}
     * @param accountNo Account Number which needs to searched in {@link List} of
     * {@link AccountOption}
     * @return Returns {@link AccountOption} which has Account Number same as the provided
     * {@code accountNo} in function parameter.
     */
    public AccountOption searchAccount(List<AccountOption> accountOptions, final String accountNo) {
        final AccountOption[] account = {new AccountOption()};
        Observable.fromIterable(accountOptions)
                .filter(new Predicate<AccountOption>() {
                    @Override
                    public boolean test(AccountOption accountOption) {
                        return accountOption.getAccountNo().equals(accountNo);
                    }
                })
                .subscribe(new Consumer<AccountOption>() {
                    @Override
                    public void accept(AccountOption accountOption) throws Exception {
                        account[0] = accountOption;
                    }
                });
        return account[0];
    }
}

File: app/src/main/java/org/mifos/mobile/presenters/ThirdPartyTransferPresenter.kt
package org.mifos.mobile.presenters

import android.content.Context

import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.BiFunction
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers

import org.mifos.mobile.R
import org.mifos.mobile.api.DataManager
import org.mifos.mobile.injection.ApplicationContext
import org.mifos.mobile.models.AccountOptionAndBeneficiary
import org.mifos.mobile.models.beneficiary.Beneficiary
import org.mifos.mobile.models.beneficiary.BeneficiaryDetail
import org.mifos.mobile.models.payload.AccountDetail
import org.mifos.mobile.models.templates.account.AccountOption
import org.mifos.mobile.models.templates.account.AccountOptionsTemplate
import org.mifos.mobile.presenters.base.BasePresenter
import org.mifos.mobile.ui.views.ThirdPartyTransferView

import java.util.*
import javax.inject.Inject

/**
 * Created by dilpreet on 21/6/17.
 */
class ThirdPartyTransferPresenter @Inject constructor(
        private val dataManager: DataManager?,
        @ApplicationContext context: Context?
) : BasePresenter<ThirdPartyTransferView?>(context) {
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
    override fun attachView(mvpView: ThirdPartyTransferView?) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        compositeDisposable.clear()
    }

    /**
     * Fetches [AccountOptionsTemplate] and [List] of [Beneficiary] from server
     * and notifies the view to display them. And in case of any error during fetching the required
     * details it notifies the view.
     */
    fun loadTransferTemplate() {
        checkViewAttached()
        mvpView?.showProgress()
        compositeDisposable.add(Observable.zip(dataManager?.thirdPartyTransferTemplate,
                dataManager?.beneficiaryList,
                BiFunction<AccountOptionsTemplate?, List<Beneficiary?>?, AccountOptionAndBeneficiary> { accountOptionsTemplate, beneficiaries ->
                    AccountOptionAndBeneficiary(accountOptionsTemplate,
                            beneficiaries)
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(object : DisposableObserver<AccountOptionAndBeneficiary?>() {
                    override fun onComplete() {}
                    override fun onError(e: Throwable) {
                        mvpView?.hideProgress()
                        mvpView?.showError(context?.getString(
                                R.string.error_fetching_third_party_transfer_template))
                    }

                    override fun onNext(accountOptionAndBeneficiary: AccountOptionAndBeneficiary) {
                        mvpView?.hideProgress()
                        mvpView?.showThirdPartyTransferTemplate(accountOptionAndBeneficiary.accountOptionsTemplate)
                        mvpView?.showBeneficiaryList(accountOptionAndBeneficiary.beneficiaryList)
                    }
                }))
    }

    /**
     * Retrieving [List] of `accountNumbers` from [List] of [AccountOption]
     *
     * @param accountOptions [List] of [AccountOption]
     * @return Returns [List] containing `accountNumbers`
     */
    fun getAccountNumbersFromAccountOptions(accountOptions: List<AccountOption>?): List<AccountDetail> {
        val accountNumber: MutableList<AccountDetail> = ArrayList()
        Observable.fromIterable(accountOptions)
                .filter { (_, _, accountType) -> accountType?.code != context?.getString(R.string.account_type_loan) }
                .flatMap { (_, accountNo, accountType) ->
                    Observable.just(AccountDetail(accountNo!!,
                            accountType?.value!!))
                }
                .subscribe { accountDetail -> accountNumber.add(accountDetail) }
        return accountNumber
    }

    /**
     * Retrieving [List] of `accountNumbers` from [List] of [Beneficiary]
     *
     * @param beneficiaries [List] of [Beneficiary]
     * @return Returns [List] containing `accountNumbers`
     */
    fun getAccountNumbersFromBeneficiaries(beneficiaries: List<Beneficiary?>?): List<BeneficiaryDetail> {
        val accountNumbers: MutableList<BeneficiaryDetail> = ArrayList()
        Observable.fromIterable(beneficiaries)
                .flatMap { (_, name, _, _, _, accountNumber) ->
                    Observable.just(BeneficiaryDetail(accountNumber,
                            name))
                }
                .subscribe { beneficiaryDetail -> accountNumbers.add(beneficiaryDetail) }
        return accountNumbers
    }

    /**
     * Searches for a [AccountOption] with provided `accountNo` from [List] of
     * [AccountOption] and returns it.
     *
     * @param accountOptions [List] of [AccountOption]
     * @param accountNo      Account Number which needs to searched in [List] of
     * [AccountOption]
     * @return Returns [AccountOption] which has Account Number same as the provided
     * `accountNo` in function parameter.
     */
    fun searchAccount(accountOptions: List<AccountOption>?, accountNo: String?): AccountOption {
        val account = arrayOf(AccountOption())
        Observable.fromIterable(accountOptions)
                .filter { (_, accountNo1) -> accountNo1 == accountNo }
                .subscribe { accountOption -> account[0] = accountOption }
        return account[0]
    }

}